File: C++/Tree/1991.cpp
#include <iostream>

using namespace std;

struct Node {
	int left;
	int right;
};

Node a[50];

void preorder(int x) {
	if (x == -1) return;
	cout << (char)(x + 'A');	//root
	preorder(a[x].left);		// 왼
	preorder(a[x].right);		// 오
}

void inorder(int x) {
	if (x == -1) return;
	inorder(a[x].left);		//왼
	cout << (char)(x + 'A');	//root
	inorder(a[x].right);		// 오
}

void postorder(int x) {
	if (x == -1) return;
	postorder(a[x].left);		// 왼
	postorder(a[x].right);		// 오
	cout << (char)(x + 'A');	// root
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	int n;
	cin >> n;

	for (int i = 1; i <= n; i++) {

		char x, y, z;
		cin >> x >> y >> z;
		x = x - 'A';

		if (y == '.') {
			a[x].left = -1;
		}
		else {
			a[x].left = y - 'A';
		}
		if (z == '.') {
			a[x].right = -1;
		}
		else {
			a[x].right = z - 'A';
		}
	}

	preorder(0);
	cout << '\n';

	inorder(0);
	cout << '\n';

	postorder(0);
	cout << '\n';

	return 0;
}

File: Python/구현/1991.py
# https://www.acmicpc.net/problem/1991
import sys


class Node:
    def __init__(self, left, right):
        self.left = left
        self.right = right

# 전위
def preorder(target: int):
    if target == -1:
        return

    print(chr(target + ord("A")), end="")
    preorder(nodes[target].left)
    preorder(nodes[target].right)


# 중위
def inorder(target: int):
    if target == -1:
        return

    inorder(nodes[target].left)
    print(chr(target + ord("A")), end="")
    inorder(nodes[target].right)

# 후위
def postorder(target: int):
    if target == -1:
        return

    postorder(nodes[target].left)
    postorder(nodes[target].right)
    print(chr(target + ord("A")), end="")


n = int(input())

# root의 알파벳 아스키코드값을 idx로 쓸거임
# [TODO] 이렇게 안하고 딕셔너리 사용해서 풀어도 될듯!! 그러면 아스키코드값도 안써도 될듯
nodes = [None for _ in range(65)]  

for i in range(n):
    root, left, right = input().rstrip().split()

    root = ord(root) - ord("A")

    if left == ".":
        left = -1
    else:
        left = ord(left) - ord("A")

    if right == ".":
        right = -1
    else:
        right = ord(right) - ord("A")

    node = Node(left, right)
    nodes[root] = node

# A가 항상 루트이므로 0 넣어주면 됨
preorder(0)
print("")

inorder(0)
print("")

postorder(0)
